### PR TITLE
fix(protocol): Properly document which timestamps are accepted

### PR DESCRIPTION
**Pull request by untitaker**
Also remove use of a `FromStr` impl where chrono's documentation says
nothing about the format that is being parsed.

It is currently unclear to me whether the timestamp format we accept in prod is even all of RFC 3339.

### CODE REVIEW

File: relay-general/src/protocol/types.rs [hunk 1]
```diff
@@ -976,9 +976,9 @@ impl FromValue for Timestamp {
     fn from_value(value: Annotated<Value>) -> Annotated<Self> {
         let rv = match value {
             Annotated(Some(Value::String(value)), mut meta) => {
-                let parsed = match value.parse::<NaiveDateTime>() {
+                let parsed = match NaiveDateTime::parse_from_str(&value, "%Y-%m-%dT%H:%M:%S%.f") {
```
Comment (jjbayer) [inline relay-general/src/protocol/types.rs]:
```suggestion
                // Use the same format as `NaiveDateTime::FromStr`: https://docs.rs/chrono/0.4.19/chrono/naive/struct.NaiveDateTime.html#impl-FromStr
                let parsed = match NaiveDateTime::parse_from_str(&value, "%Y-%m-%dT%H:%M:%S%.f") {
```

File: relay-general/src/protocol/types.rs [hunk 2]
```diff
@@ -976,9 +976,9 @@ impl FromValue for Timestamp {
...
                     Ok(dt) => Ok(DateTime::from_utc(dt, Utc)),
-                    Err(_) => value.parse(),
+                    Err(_) => DateTime::parse_from_rfc3339(&value).map(|dt| dt.with_timezone(&Utc)),
```
Comment (jjbayer) [inline relay-general/src/protocol/types.rs]: Is this a stricter parser, or is it exactly the same?

Comment (untitaker) [reply relay-general/src/protocol/types.rs]: It's a more lenient parser afaict, but chrono literally does not document what `FromStr` actually accepts. So, not sure. Will sit on this PR for a bit.

**Review by jjbayer — APPROVED** [review]

File: relay-general/src/protocol/types.rs [hunk 3]
```diff
@@ -976,8 +976,18 @@ impl FromValue for Timestamp {
...
             Annotated(Some(Value::String(value)), mut meta) => {
-                let parsed = match value.parse::<NaiveDateTime>() {
+                // Use the same format as `NaiveDateTime::FromStr`, just fixed so that it won't
+                // change between library upgrades:
+                // https://docs.rs/chrono/0.4.19/chrono/naive/struct.NaiveDateTime.html#impl-FromStr
+                let parsed = match NaiveDateTime::parse_from_str(&value, "%Y-%m-%dT%H:%M:%S%.f") {
                     Ok(dt) => Ok(DateTime::from_utc(dt, Utc)),
+
+                    // XXX: This actually accepts more than RFC 3339. SDKs are strongly discouraged
+                    // from exercising that freedom. We should only support RFC3339.
```
Comment (jan-auer) [inline relay-general/src/protocol/types.rs:984]: The implementation suggestion for SDKs is in a very hidden spot. If we keep current behavior, I would suggest removing the comment here and moving it to a doc comment or develop docs.

File: relay-general/src/protocol/types.rs (hunk 1 above)
Comment (jan-auer) [reply relay-general/src/protocol/types.rs]: This is lenient against spaces, but it's not obvious from the documentation or the format string. Can you please add a test case for something like `"2015-  09-18T23:56:04"`?

File: relay-general/src/protocol/event.rs [hunk 4]
```diff
@@ -304,14 +304,23 @@ pub struct Event {
     /// value representing the number of seconds that have elapsed since the [Unix
     /// epoch](https://en.wikipedia.org/wiki/Unix_time).
     ///
+    /// Timezone is assumed to be UTC if unclear.
```
Comment (jan-auer) [inline relay-general/src/protocol/event.rs]:
```suggestion
    /// Timezone is assumed to be UTC if missing.
```

File: relay-general/src/protocol/event.rs [hunk 5]
```diff
@@ -304,14 +304,23 @@ pub struct Event {
...
+    ///
     /// Sub-microsecond precision is not preserved with numeric values due to precision
     /// limitations with floats (at least in our systems). With that caveat in mind, just send
     /// whatever is easiest to produce.
     ///
-    /// All timestamps in the event protocol are formatted this way.
+    /// All timestamps in the event protocol, be it errors, transactions or the spans within a
```
Comment (jan-auer) [inline relay-general/src/protocol/event.rs]: I would revert this sentence, it reads strange and doesn't add any insight.

**Review by jan-auer — CHANGES_REQUESTED** [review]
I'm +1 on the documentation changes.

However, I'm conflicted whether the explicit date time format is better than using the built-in `FromStr`, which is both documented and doesn't require parsing a format string over and over again. I would suggest that if we're touching the implementation, we should rather look into parsing full ISO 8601 in addition to RFC 3339 and then add sufficient tests.

File: relay-general/src/protocol/types.rs (hunk 3 above)
Comment (untitaker) [reply relay-general/src/protocol/types.rs:984]: The "real" docs already tell the SDKs to use RFC 3339, this code comment is intended to explain why the actual relay code accepts more.

File: relay-general/src/protocol/types.rs (hunk 1 above)
Comment (untitaker) [reply relay-general/src/protocol/types.rs]: @jan-auer is this still relevant? are you saying FromStr is lenient as well?

**Comment by untitaker** [comment]
@jan-auer I have reverted most changes and left it at adding code comments

**Review by jan-auer — APPROVED** [review]

**Pull request merged**